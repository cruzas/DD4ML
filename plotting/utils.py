# import os
# import pprint
# from itertools import product

# import matplotlib.pyplot as plt
# import numpy as np
# import pandas as pd
# from scipy.stats import median_abs_deviation

# import wandb


def _abbreviate_val(val):
    if isinstance(val, bool):
        return "T" if val else "F"
    return val
