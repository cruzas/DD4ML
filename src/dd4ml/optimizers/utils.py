def get_state_dict(model):
    return model.module.state_dict() if hasattr(model, "module") else model.state_dict()

def get_trust_region_params(config, lr_scale=1.0, max_iter=3):
    return {
        "lr": config.learning_rate * lr_scale,
        "max_lr": 2.0,
        "min_lr": 1e-4,
        "nu": 0.5,
        "inc_factor": 2.0,
        "dec_factor": 0.5,
        "nu_1": 0.25,
        "nu_2": 0.75,
        "max_iter": max_iter,
        "norm_type": 2,
    }
    
def get_local_trust_region_params(config, lr_scale=1.0, max_iter=3):
    return {
        "lr": config.learning_rate * lr_scale,
        "max_lr": 1.5,        # Lower maximum for local updates
        "min_lr": 1e-4,
        "nu": 0.45,           # Adjusted to be more conservative locally
        "inc_factor": 1.5,    # Reduced increase factor
        "dec_factor": 0.6,    # Slightly more aggressive reduction
        "nu_1": 0.3,
        "nu_2": 0.7,
        "max_iter": max_iter,
        "norm_type": 2,
    }

