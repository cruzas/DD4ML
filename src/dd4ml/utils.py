import copy
import importlib
import pprint

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
from torch.nn.parallel import DistributedDataParallel as DDP

import wandb
from dd4ml.utility.dist_utils import *
from dd4ml.utility.mingpt_utils import *
from dd4ml.utility.ml_utils import *

# Check whether wandb exists
try:
    from dd4ml.utility.wandb_utils import *
except ImportError:
    pass


# Utility: Dynamically import an attribute from a module.
def import_attr(module_path: str, attr: str):
    module = importlib.import_module(module_path)
    return getattr(module, attr)


# Utility: Broadcast a dictionary using PyTorch's distributed utilities.
def broadcast_dict(d, src=0):
    obj_list = [d] if dist.get_rank() == src else [None]
    dist.broadcast_object_list(obj_list, src=src)
    return obj_list[0]


# Generic Factory for creating components.
class Factory:
    def __init__(self, mapping: dict):
        self.mapping = mapping

    def create(self, key: str, *args, **kwargs):
        if key not in self.mapping:
            raise ValueError(f"Unknown key: {key}")
        module_path, creator = self.mapping[key]
        if callable(creator):
            return creator(*args, **kwargs)
        cls = import_attr(module_path, creator)
        return cls(*args, **kwargs)


# Mapping definitions.
DATASET_MAP = {
    "mnist": ("dd4ml.datasets.mnist", "MNISTDataset"),
    "cifar10": ("dd4ml.datasets.cifar10", "CIFAR10Dataset"),
    "tinyshakespeare": ("dd4ml.datasets.tinyshakespeare", "TinyShakespeareDataset"),
}

MODEL_MAP = {
    "simple_cnn": ("dd4ml.models.cnn.simple_cnn", "SimpleCNN"),
    "big_cnn": ("dd4ml.models.cnn.big_cnn", "BigCNN"),
    "simple_resnet": ("dd4ml.models.resnet.simple_resnet", "SimpleResNet"),
    "mingpt": ("dd4ml.models.gpt.mingpt.model", "GPT"),
}

CRITERION_MAP = {
    "cross_entropy": ("", lambda ds=None: nn.CrossEntropyLoss()),
    "weighted_cross_entropy": (
        "",
        lambda ds: nn.CrossEntropyLoss(weight=ds.compute_class_weights()),
    ),
    "mse": ("", lambda ds=None: nn.MSELoss()),
    "cross_entropy_transformers": (
        "",
        lambda ds=None: cross_entropy_transformers,
    ),  # Assumes defined elsewhere.
}

OPTIMIZER_MAP = {
    "sgd": ("", lambda model, lr: optim.SGD(model.parameters(), lr=lr, momentum=0.9)),
    "adam": ("", lambda model, lr: optim.Adam(model.parameters(), lr=lr)),
    "adamw": ("", lambda model, lr: optim.AdamW(model.parameters(), lr=lr)),
    "adagrad": ("", lambda model, lr: optim.Adagrad(model.parameters(), lr=lr)),
    "rmsprop": ("", lambda model, lr: optim.RMSprop(model.parameters(), lr=lr)),
}

# Instantiate factories.
dataset_factory = Factory(DATASET_MAP)
model_factory = Factory(MODEL_MAP)
criterion_factory = Factory(CRITERION_MAP)
optimizer_factory = Factory(OPTIMIZER_MAP)


# Configuration cleanup: recursively remove keys.
def remove_keys(config, keys_to_remove):
    if isinstance(config, dict):
        for k in keys_to_remove:
            config.pop(k, None)
        for key, value in config.items():
            config[key] = remove_keys(value, keys_to_remove)
    elif hasattr(config, "__dict__"):
        for k in keys_to_remove:
            if k in config.__dict__:
                del config.__dict__[k]
        for k in list(config.__dict__.keys()):
            value = getattr(config, k)
            updated_value = remove_keys(value, keys_to_remove)
            if hasattr(config, k):  # Check if the attribute still exists
                setattr(config, k, updated_value)
    return config


# Standardize configuration by removing unnecessary keys.
def make_std_config(config):
    use_pmw = getattr(config.trainer, "use_pmw", False)
    if not use_pmw:
        keys_to_remove = ["num_stages", "num_replicas_per_subdomain", "model_handler"]
        if "apts_d" not in config.optimizer.lower():
            keys_to_remove.append("num_subdomains")
        config = remove_keys(config, keys_to_remove)
    if config.optimizer != "apts":
        keys_to_remove = [
            "subdomain_optimizer",
            "subdomain_optimizer_args",
            "global_optimizer",
            "global_optimizer_args",
        ]
        config = remove_keys(config, keys_to_remove)
    return config


# Refactored configuration creation.
def get_config(dataset_name: str, model_name: str, optimizer: str):
    from dd4ml.trainer import (
        Trainer,
    )  # Assumes Trainer and CfgNode are defined in dd4ml.trainer.

    C = CfgNode()  # CfgNode should be defined elsewhere.

    # System configuration.
    C.system = CfgNode()
    C.system.seed = 3407
    C.system.trial = 0
    C.system.work_dir = f"../../saved_networks/{dataset_name}/{model_name}/{optimizer}/"

    # Data configuration via DATASET_MAP.
    if dataset_name not in DATASET_MAP:
        raise ValueError(f"Unknown dataset name: {dataset_name}")
    ds_module, ds_class_name = DATASET_MAP[dataset_name]
    dataset_cls = import_attr(ds_module, ds_class_name)
    C.data = dataset_cls.get_default_config()

    # Model configuration via MODEL_MAP.
    key = next((k for k in MODEL_MAP if k in model_name.lower()), None)
    if key is None:
        raise ValueError(f"Unknown model name: {model_name}.")
    model_module, model_class_name = MODEL_MAP[key]
    model_cls = import_attr(model_module, model_class_name)
    C.model = model_cls.get_default_config()
    C.model.model_class = model_cls

    # Set image-specific properties if available.
    for attr in ["input_channels", "input_height", "input_width", "output_classes"]:
        if hasattr(C.data, attr):
            setattr(C.model, attr, getattr(C.data, attr))

    # Trainer configuration.
    C.trainer = Trainer.get_default_config()
    return C


# Refactored function to create config, model, and trainer.
def get_config_model_and_trainer(args, wandb_config):
    from dd4ml.pmw.model_handler import ModelHandler
    from dd4ml.pmw.parallelized_model import ParallelizedModel
    from dd4ml.trainer import Trainer

    config_src = wandb_config if wandb_config is not None else args
    dataset_name = config_src["dataset_name"]
    model_name = config_src["model_name"]
    optimizer_name = config_src["optimizer"]

    all_config = get_config(dataset_name, model_name, optimizer_name)
    if wandb_config is not None:
        all_config.merge_from_dict(wandb_config)
    else:
        args.pop("sweep_config", None)
    all_config.merge_from_dict(args)
    all_config.merge_and_cleanup(keys_to_look=["system", "model", "trainer"])

    # Dataset instantiation via factory.
    dataset = dataset_factory.create(all_config.dataset_name, all_config.data)
    test_dataset_config = copy.deepcopy(all_config.data)
    test_dataset_config.train = False
    test_dataset = dataset.__class__(test_dataset_config)

    # For GPT or similar models.
    if hasattr(dataset, "vocab_size"):
        all_config.model.vocab_size = dataset.get_vocab_size()
    if hasattr(dataset, "block_size"):
        all_config.model.block_size = dataset.get_block_size()

    all_config = make_std_config(all_config)

    # Model instantiation and optional parallelization.
    if getattr(all_config.trainer, "use_pmw", False) and hasattr(
        all_config.model.model_class, "as_model_dict"
    ):
        dprint("Using Parallel Model Wrapper and Model Handler")
        model_instance = all_config.model.model_class(all_config.model)
        model_dict = model_instance.as_model_dict()
        model_handler = ModelHandler(
            model_dict,
            all_config.model.num_subdomains,
            all_config.model.num_replicas_per_subdomain,
        )
        if dist.get_rank() == 0:
            pprint.pprint(model_handler.nn_structure)
        all_config.trainer.model_handler = model_handler

        sample_input = dataset.get_sample_input(all_config.trainer)
        if sample_input.shape[0] == 1:
            other_sample = dataset.get_sample_input(all_config.trainer)
            sample_input = torch.cat([sample_input, other_sample], dim=0)
            
        model = ParallelizedModel(model_handler, sample=sample_input)
    else:
        model = all_config.model.model_class(all_config.model)

    dprint(model)

    # Criterion selection.
    criterion_key = (
        wandb_config["criterion"] if wandb_config is not None else args["criterion"]
    )
    if criterion_key not in CRITERION_MAP:
        raise ValueError(f"Unknown criterion: {criterion_key}")
    criterion = criterion_factory.create(
        criterion_key, dataset if "weighted" in criterion_key else None
    )

    # Optimizer selection.
    lr = (
        wandb_config["learning_rate"]
        if wandb_config is not None
        else args["learning_rate"]
    )
    if optimizer_name in OPTIMIZER_MAP:
        optimizer_obj = optimizer_factory.create(optimizer_name, model, lr)
        for attr in [
            "subdomain_optimizer",
            "subdomain_optimizer_args",
            "global_optimizer",
            "global_optimizer_args",
        ]:
            if hasattr(all_config.trainer, attr):
                delattr(all_config.trainer, attr)
    elif optimizer_name == "trust_region":
        from dd4ml.optimizers.trust_region import TrustRegion

        all_config.trainer = TrustRegion.setup_TR_args(all_config.trainer)
        optimizer_obj = TrustRegion(
            model=model,
            lr=all_config.trainer.learning_rate,
            max_lr=all_config.trainer.max_lr,
            min_lr=all_config.trainer.min_lr,
            nu=all_config.trainer.nu,
            inc_factor=all_config.trainer.inc_factor,
            dec_factor=all_config.trainer.dec_factor,
            nu_1=all_config.trainer.nu_1,
            nu_2=all_config.trainer.nu_2,
            max_iter=all_config.trainer.max_iter,
            norm_type=all_config.trainer.norm_type,
        )

    elif optimizer_name == "apts":
        from dd4ml.optimizers.apts import APTS

        all_config.trainer = APTS.setup_APTS_args(all_config.trainer)
        optimizer_obj = APTS(
            model=model,
            subdomain_optimizer=all_config.trainer.subdomain_optimizer,
            subdomain_optimizer_defaults=all_config.trainer.subdomain_optimizer_args,
            global_optimizer=all_config.trainer.global_optimizer,
            global_optimizer_defaults=all_config.trainer.global_optimizer_args,
            lr=all_config.trainer.learning_rate,
            max_subdomain_iter=all_config.trainer.max_subdomain_iters,
            dogleg=True,
            APTS_in_data_sync_strategy="average",
            step_strategy="mean",
        )
    elif optimizer_name == "apts_d":
        from dd4ml.optimizers.apts_d import APTS_D

        all_config.trainer = APTS_D.setup_APTS_args(all_config.trainer)
        all_config.trainer.apts_d = True
        local_rank = int(os.environ.get("LOCAL_RANK", 0))
        device = (
            f"cuda:{torch.cuda.current_device()}"
            if dist.get_backend() != "gloo"
            else "cpu"
        )
        model.to(device)
        model = DDP(
            model, device_ids=[local_rank] if torch.cuda.is_available() else None
        )
        optimizer_obj = APTS_D(
            params=model.parameters(),
            model=model,
            criterion=criterion,
            device=device,
            max_iter=3,
            nr_models=all_config.model.num_subdomains,
            global_opt=all_config.trainer.global_optimizer,
            global_opt_params=all_config.trainer.global_optimizer_args,
            local_opt=all_config.trainer.subdomain_optimizer,
            local_opt_params=all_config.trainer.subdomain_optimizer_args,
            global_pass=True,
            foc=True,
        )
    else:
        raise ValueError(f"Unknown optimizer: {optimizer_name}")

    # Set training schedule based on dataset.
    if "shakespeare" in all_config.dataset_name:
        all_config.trainer.run_by_epoch = False
    else:
        all_config.trainer.run_by_epoch = True

    trainer = Trainer(
        all_config.trainer, model, optimizer_obj, criterion, dataset, test_dataset
    )
    return all_config, model, trainer


# Entry point for running the experiment.
def generic_run(
    rank=None,
    args=None,
    wandb_config=None,
    epoch_end_callback=None,
    batch_end_callback=None,
):

    use_wandb = wandb_config is not None
    if use_wandb:
        if rank == 0:  # Only rank 0 initializes wandb.
            if wandb.run is None:
                wandb.init(config=wandb_config)
            wandb_config = dict(wandb.config)
        else:
            wandb_config = {}
        # Ensure all ranks use the same hyperparameters.
        wandb_config = broadcast_dict(wandb_config, src=0)
    else:
        wandb_config = {}

    if "apts_d" in wandb_config["optimizer"].lower():
        args["use_pmw"] = False
        args["num_subdomains"] = dist.get_world_size() if dist.is_initialized() else 1

    config, _, trainer = get_config_model_and_trainer(args, wandb_config)
    dprint(config)

    if epoch_end_callback and trainer.config.run_by_epoch:
        trainer.set_callback("on_epoch_end", epoch_end_callback)
    if batch_end_callback and not trainer.config.run_by_epoch:
        trainer.set_callback("on_batch_end", batch_end_callback)

    trainer.run()
