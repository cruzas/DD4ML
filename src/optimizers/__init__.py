# __init__.py

# Manually import specific modules or functions to be included
# in wildcard imports
from .APTS import APTS
from .TR import TR
from .LocalTR import LocalTR
from .TRAdam import TRAdam

# Define __all__ to specify what should be imported when using *
__all__ = ['APTS', 'TR', 'LocalTR', 'TRAdam']
