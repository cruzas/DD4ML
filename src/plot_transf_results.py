import matplotlib.pyplot as plt
import pandas as pd

# Load the data files
files = {
    "SGD, nsd=2": "../results/tshakespeare_t_0_nsd_2_nrs_1_nst_13_bs_2048_parsgd_iter.csv",
    "APTS, nsd=2": "../results/tshakespeare_t_0_nsd_2_nrs_1_nst_13_bs_2048_iter.csv",
}

# Define plot styles
styles = {
    "SGD, nsd=2": {"color": "blue", "linestyle": "-"},
    "APTS, nsd=2": {"color": "blue", "linestyle": "--"},
}

# Plot the data
plt.figure(figsize=(10, 6))
for label, filepath in files.items():
    data = pd.read_csv(filepath)
    plt.plot(data["iteration"], data["loss"], label=label, **styles[label])

# Customize the plot
plt.xlabel("Iteration")
plt.ylabel("Loss")
plt.title("Loss vs Iteration")
plt.legend()
plt.grid(True)

# Show the plot
plt.show()
