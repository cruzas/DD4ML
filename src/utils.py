import os  
import torch
import subprocess
import torch.distributed as dist


def decide_tensor_device(ws, backend, gpu_id):
    if torch.cuda.is_available():
        if backend == 'gloo':
            if torch.cuda.device_count() < ws:
                return f'cuda:{gpu_id}'
            else:
                return f'cuda:{dist.get_rank()}'
        else:
            return f'cuda:{gpu_id}'
    else:
        return 'cpu'

def find_free_port():
    """
    References:
        - https://stackoverflow.com/questions/1365265/on-localhost-how-do-i-pick-a-free-port-number
    """
    import socket
    from contextlib import closing
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        s.bind(('', 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return str(s.getsockname()[1])  

def prepare_distributed_environment(rank=None, master_addr=None, master_port=None, world_size=None, is_cuda_enabled=True):
    if not is_cuda_enabled:
        os.environ["CUDA_VISIBLE_DEVICES"]="" # TODO: Remove this line. It's just for debugging purposes.
    if rank is None and master_addr is None and master_port is None and world_size is None: # we are on a cluster
        print(f'Should be initializing {os.environ["SLURM_NNODES"]} nodes')
        ## Execute code on a cluster
        os.environ['MASTER_PORT'] = '29501'
        os.environ['WORLD_SIZE'] = os.environ['SLURM_NNODES']
        os.environ['LOCAL_RANK'] = '0'
        os.environ['RANK'] = os.environ['SLURM_NODEID']
        node_list = os.environ['SLURM_NODELIST']
        master_node = subprocess.getoutput(
            f'scontrol show hostname {node_list} | head -n1'
        )
        os.environ['MASTER_ADDR'] = master_node
        dist.init_process_group(backend='nccl')
    else: # To execute on a PC
        os.environ['MASTER_ADDR'] = master_addr
        os.environ['MASTER_PORT'] = master_port if master_port is not None else find_free_port()
        dist.init_process_group(backend='gloo', rank=rank, world_size=world_size)
        
def send_shape(shape: list, dst: int, device = None):
    if device is None:
        device = torch.device('cuda') if dist.get_backend() == 'nccl' else torch.device('cpu')
    for s in shape:
        dist.send(tensor=torch.tensor(s, dtype=torch.int32).to(device), dst=dst)
    dist.send(tensor=torch.tensor(-1, dtype=torch.int32).to(device), dst=dst)

def receive_shape(src: int, device = None):
    # Rest of the code...
    if device is None:
        device = torch.device('cuda') if dist.get_backend() == 'nccl' else torch.device('cpu')
    shape = []; temp = 0
    while True:
        temp = torch.tensor((0), dtype=torch.int32).to(device)
        dist.recv(tensor=temp, src=src)
        if temp == -1:
            break
        shape.append(temp.item())
    return shape

def closure(inputs, targets, criterion, model, compute_grad=True, zero_grad=True, return_output=False, data_chunks_amount=1):
    """
    NOTE: Losses from different chunks are averaged.
    """
    # if model.__class__.__name__ != 'ParallelizedModel':
    #     raise ValueError('Model must be an instance of the "ParallelizedModel".')
    if isinstance(criterion, type):
        raise ValueError('Criterion must be an instance of a class.')
    if model.rank in model.all_stage_ranks[-1]:
        targets = targets.chunk(data_chunks_amount)
    # Compute loss
    def closure2(compute_grad=compute_grad, zero_grad=zero_grad, data_chunks_amount=data_chunks_amount):
        if zero_grad:
            model.zero_grad()
        with torch.set_grad_enabled(compute_grad):
            outputs = model(inputs, chunks_amount=data_chunks_amount)
        # loss = torch.zeros(1).to(model.gpu_device)
        losses = [0] * data_chunks_amount
        loss = torch.tensor(0.0).to(model.tensor_device)
        if model.rank in model.all_stage_ranks[-1]:
            for i,out in enumerate(outputs):
                losses[i] = criterion(out, targets[i].to(out.device))
            loss = sum(losses)/len(losses)
        # Average losses across replicas
        if model.rank in model.all_stage_ranks[-1]:
            dist.all_reduce(tensor=loss, op=dist.ReduceOp.SUM, group=model.layer_copies_group) # Summing the losses across final layers of each replicas
            loss = loss/model.tot_replicas
        loss_broadcast = dist.broadcast(tensor=loss.detach(), src=model.all_stage_ranks[-1][0], async_op=True) # Last layer of first model replica broadcasts the loss to all other ranks 
        if compute_grad and torch.is_grad_enabled():
            model.backward(losses)
        loss_broadcast.wait()
        if return_output:
            if model.rank in model.all_stage_ranks[-1]:
                # Returning outputs here in case we want to compute the accuracy afterwards
                return loss.item(), [output for output in outputs]
            else:
                return loss.item(), None
        return loss.item()
    return closure2 

def check_gpus_per_rank():
    '''
    Ensure that the number of GPUs is the same on every rank in the distributed environment.
    This is necessary to perform tensor sharding. 
    '''
    # Get the number of GPUs available on the current rank
    local_gpus = torch.cuda.device_count()

    # Gather the number of GPUs from all ranks
    gpu_counts = [torch.tensor(0).cuda() for _ in range(dist.get_world_size())]
    dist.all_gather(gpu_counts, torch.tensor(local_gpus).cuda())

    # Convert gathered tensors to CPU and list
    gpu_counts = [gpu.item() for gpu in gpu_counts]

    # Check if all ranks have the same number of GPUs
    if len(set(gpu_counts)) != 1:
        raise ValueError("Mismatch in the number of GPUs across ranks")