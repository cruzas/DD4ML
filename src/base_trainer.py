"""
Simple training loop; Boilerplate that could apply to any arbitrary neural network,
so nothing in this file really has anything to do with GPT specifically.
"""

import time
from abc import ABC, abstractmethod
from collections import defaultdict

import torch
from torch.utils.data.dataloader import DataLoader

from src.utils import CfgNode as CN
from src.utils import closure, dprint


class BaseTrainer(ABC):
    @staticmethod
    def get_default_config():
        C = CN()
        # device to train on
        C.device = 'auto'
        # dataloder parameters
        C.num_workers = 1
        # optimizer parameters
        C.max_iters = None
        C.batch_size = 64
        C.learning_rate = 3e-4
        C.betas = (0.9, 0.95)
        C.weight_decay = 0.1  # only applied on matmul weights
        C.grad_norm_clip = 1.0 
        C.epochs = 1 # in case epochs instead of iter
        return C

    def __init__(self, config, model, optimizer, criterion, train_dataset, test_dataset):
        self.config = config
        self.model = model
        self.optimizer = optimizer
        self.criterion = criterion
        self.train_dataset = train_dataset
        self.test_dataset = test_dataset
        self.callbacks = defaultdict(list)

        # determine the device we'll train on
        if config.device == 'auto':
            self.device = 'cuda' if torch.cuda.is_available() else 'cpu'
        else:
            self.device = config.device
        self.model = self.model.to(self.device)
        dprint(f"running on device {self.device}")

        # variables that will be assigned to trainer class later for logging and etc
        self.iter_num = 0
        self.iter_time = 0.0
        self.iter_dt = 0.0

    def add_callback(self, onevent: str, callback):
        self.callbacks[onevent].append(callback)

    def set_callback(self, onevent: str, callback):
        self.callbacks[onevent] = [callback]

    def trigger_callbacks(self, onevent: str):
        for callback in self.callbacks.get(onevent, []):
            callback(self)
    
    def compute_epoch_loss(self):
        model.train()
        model, config = self.model, self.config
        criterion = self.criterion
        self.loss = 0.0 # epoch loss
        for batch_idx, (x, y) in enumerate(self.train_loader):    
            x, y = x.to(self.device), y.to(self.device)
            
            if self.epoch_num == 0:
                first_closure = closure(x, y, criterion, model, data_chunks_amount=config.data_chunks_amount, compute_grad=False)
                self.loss += first_closure()
            else:
                general_closure = closure(x, y, criterion=criterion, model=model, data_chunks_amount=config.data_chunks_amount, compute_grad=True, grad_norm_clip=config.grad_norm_clip)   
                
                # Check if final_subdomain_closure is part of self.optimizer arguments
                if hasattr(self.optimizer, 'final_subdomain_closure'):
                    def final_subdomain_closure(outputs, y=y):
                        y_chunks = y.chunk(len(outputs))
                        loss = []
                        for i, o in enumerate(outputs):
                            loss.append(self.criterion(o, y_chunks[i]))
                        return loss
                
                    self.loss += self.optimizer.step(closure=general_closure, final_subdomain_closure=final_subdomain_closure)
                else:
                    self.loss += self.optimizer.step(closure)
                
            # Print progress within the epoch
            self.epoch_progress = 100.0 * (batch_idx + 1) / total_batches
            self.trigger_callbacks('on_batch_end')
        
        self.loss = self.loss / total_batches
        tnow = time.time()
        self.epoch_dt = tnow - self.epoch_time
        self.epoch_time = tnow
    
    def compute_accuracy(self):
        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for batch_idx, (x, y) in enumerate(test_loader):
                x, y = x.to(self.device), y.to(self.device)
                outputs = model(x)
                _, predicted = torch.max(outputs, 1)
                total += y.size(0)
                correct += (predicted == y).sum().item()
            
        self.accuracy =  100.0 * correct / total
    
    def run_by_epoch(self):
        model, config = self.model, self.config
        train_loader = self.train_loader
        
        self.total_start_time = time.time()
        self.epoch_num = 0
        self.epoch_time = time.time()
        total_batches = len(train_loader)
        for epoch in range(config.epochs+1):
            self.compute_epoch_loss()
            self.compute_accuracy()
            self.running_time = time.time() - self.total_start_time
            self.trigger_callbacks('on_epoch_end')
            self.epoch_num += 1
    
    def run_by_iter(self):
        model.train()
        self.iter_time = time.time()
        data_iter = iter(train_loader)
        
        for self.iter_num in range(config.max_iters if config.max_iters is not None else float('inf')):
            # fetch the next batch (x, y) and re-init iterator if needed
            try:
                batch = next(data_iter)
            except StopIteration:
                data_iter = iter(train_loader)
                batch = next(data_iter)
            batch = [t.to(self.device) for t in batch]
            x, y = batch
        
            if self.iter_num == 0:
                first_closure = closure(x, y, criterion, model, data_chunks_amount=config.data_chunks_amount, compute_grad=False)
                self.loss = first_closure()
            else:
                self.optimizer.zero_grad()      
                general_closure = closure(x, y, criterion=criterion, model=model, data_chunks_amount=config.data_chunks_amount, compute_grad=True, grad_norm_clip=config.grad_norm_clip)        
                self.loss = self.optimizer.step(closure=general_closure)
            
            self.trigger_callbacks('on_batch_end')
            
            tnow = time.time()
            self.iter_dt = tnow - self.iter_time
            self.iter_time = tnow

