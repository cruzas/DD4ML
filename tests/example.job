#!/bin/bash
#SBATCH --job-name=example
#SBATCH --output=./log_files/example.out
#SBATCH --error=./log_files/example.err
#SBATCH --nodes=2
#SBATCH --time=00:30:00
#SBATCH --ntasks-per-core=1
#SBATCH --ntasks-per-node=4
#SBATCH --cpus-per-task=1
#SBATCH --gres=gpu:4
#SBATCH --hint=nomultithread
#SBATCH --constraint=gpu
#SBATCH --partition=debug
#SBATCH --exclusive

# Source the environment file
source ~/.slurm_env

# Use the account name from the environment
#SBATCH --account=$SLURM_ACCOUNT

# Environment variables
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK
export NCCL_DEBUG=INFO
export NCCL_IB_HCA=ipogif0
export NCCL_IB_CUDA_SUPPORT=1

# Set the RUNPATH to the directory of the SLURM script 
RUNPATH=/capstor/scratch/cscs/scruzale/DD4ML/tests
cd $RUNPATH || exit

# Load required modules
source /opt/cray/pe/cpe/23.12/restore_lmod_system_defaults.sh
module load cray/23.12
source /opt/cray/pe/cpe/23.12/restore_lmod_system_defaults.sh
module load cray-python/3.11.5
source /opt/cray/pe/cpe/23.12/restore_lmod_system_defaults.sh

# Run the training script
srun /opt/cray/pe/python/3.11.5/bin/python -u example.py 
echo "Test done $(date)"