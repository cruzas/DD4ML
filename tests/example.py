import os

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.nn.parallel import DistributedDataParallel as DDP

from src.utils import detect_environment, prepare_distributed_environment

print(f"Cuda available: {torch.cuda.is_available()}")
print(f"Torch version: {torch.__version__}")

# Try to import wandb
try:
    import wandb
    wandb_available = True
except ImportError:
    wandb_available = False
    print("WandB is not available. Logging to stdout.")

class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2)
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(64 * 8 * 8, 256),
            nn.ReLU(),
            nn.Linear(256, 10)
        )

    def forward(self, x):
        x = self.features(x)
        return self.classifier(x)

def main(rank, master_addr, master_port, world_size, args=None):
    # Initialize the distributed environment
    prepare_distributed_environment(rank, master_addr, master_port, world_size, is_cuda_enabled=torch.cuda.is_available())
    
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    local_rank = int(os.environ['LOCAL_RANK'])
    torch.cuda.set_device(local_rank)

    print(f"Number of GPUs available: {torch.cuda.device_count()}")
    print(f"Global rank: {rank}, Local rank: {local_rank}")

    # Initialize wandb if available on rank 0
    if wandb_available and rank == 0:
        wandb.init(project="my_project", config={
            "epochs": 3,
            "batch_size": 250,
            "learning_rate": 0.01
        })

    model = SimpleCNN().cuda(local_rank)
    ddp_model = DDP(model, device_ids=[local_rank])

    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ])
    dataset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
    sampler = torch.utils.data.distributed.DistributedSampler(dataset, num_replicas=world_size, rank=rank)
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=250, sampler=sampler)

    criterion = nn.CrossEntropyLoss().cuda(local_rank)
    optimizer = optim.SGD(ddp_model.parameters(), lr=0.01)

    ddp_model.train()
    for epoch in range(3):
        sampler.set_epoch(epoch)
        epoch_loss = 0.0
        num_batches = 0
        for inputs, targets in dataloader:
            inputs = inputs.cuda(local_rank, non_blocking=True)
            targets = targets.cuda(local_rank, non_blocking=True)
            optimizer.zero_grad()
            loss = criterion(ddp_model(inputs), targets)
            loss.backward()
            optimizer.step()

            epoch_loss += loss.item()
            num_batches += 1

        # Reduce epoch loss and batch count across all processes
        epoch_loss_tensor = torch.tensor(epoch_loss, device=local_rank)
        num_batches_tensor = torch.tensor(num_batches, device=local_rank)
        dist.all_reduce(epoch_loss_tensor, op=dist.ReduceOp.SUM)
        dist.all_reduce(num_batches_tensor, op=dist.ReduceOp.SUM)
        average_loss = epoch_loss_tensor.item() / num_batches_tensor.item()

        if rank == 0:
            print(f"Epoch {epoch+1}, Average Loss: {average_loss:.4f}")
            if wandb_available:
                wandb.log({"epoch": epoch+1, "average_loss": average_loss})

    dist.destroy_process_group()

    if wandb_available and rank == 0:
        wandb.finish()

if __name__ == '__main__':
    environment = detect_environment()
    rank = None
    master_addr = None 
    master_port = None
    world_size = None
    print("Code being executed on a cluster...")
    main(rank=rank, master_addr=master_addr, master_port=master_port, world_size=world_size)