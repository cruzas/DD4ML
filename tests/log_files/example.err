Unloading the cpe module is insufficient to restore the system defaults.
Please run 'source
/opt/cray/pe/cpe/23.12/restore_lmod_system_defaults.[csh|sh]'. 
Traceback (most recent call last):
  File "/capstor/scratch/cscs/scruzale/DD4ML/tests/example.py", line 10, in <module>
    from src.utils import detect_environment, prepare_distributed_environment
  File "/capstor/scratch/cscs/scruzale/DD4ML/src/utils.py", line 7, in <module>
    import wandb
ModuleNotFoundError: No module named 'wandb'
srun: error: nid005844: task 0: Exited with exit code 1
srun: Terminating StepId=385257.0
