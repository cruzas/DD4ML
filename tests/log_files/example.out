Switching PrgEnv-cray.
Switching cce.
Switching cray-libsci.
Switching cray-mpich.
Switching cray-python.
Switching craype.
Switching perftools-base.
Switching PrgEnv-cray.
Switching cce.
Switching cray-libsci.
Switching cray-mpich.
Switching cray-python.
Switching craype.
Switching perftools-base.
Switching PrgEnv-cray.
Switching cce.
Switching cray-libsci.
Switching cray-mpich.
Switching cray-python.
Switching craype.
Switching perftools-base.
Test done Mon 10 Feb 2025 05:11:53 PM CET
